Library/Homebrew/formula_installer.rb

```diff
@@ -598,36 +596,14 @@ def expand_dependencies_for_formula(formula, inherited_options)
         Dependency.prune
       elsif dep.satisfied?(inherited_options[dep.name])
         Dependency.skip
-      else
-        any_bottle_used ||= install_bottle_for?(dep.to_formula, build)
       end
     end
-
-    [expanded_deps, any_bottle_used]
   end
 
   def expand_dependencies
     inherited_options = Hash.new { |hash, key| hash[key] = Options.new }
-    any_bottle_used = pour_bottle?
 
-    expanded_deps, any_dep_bottle_used = expand_dependencies_for_formula(formula, inherited_options)
-    any_bottle_used ||= any_dep_bottle_used
-
-    # We require some dependencies (glibc, GCC 5, etc.) if binaries were built.
-    # Native binaries shouldn't exist in cross-platform `all` bottles.
-    if any_bottle_used && !formula.bottled?(:all) && !Keg.bottle_dependencies.empty?
-      all_bottle_deps = Keg.bottle_dependencies.flat_map do |bottle_dep|
-        bottle_dep.recursive_dependencies.map(&:name) + [bottle_dep.name]
-      end
-
-      if all_bottle_deps.exclude?(formula.name)
-        bottle_deps = Keg.bottle_dependencies.flat_map do |bottle_dep|
-          expanded_bottle_deps, = expand_dependencies_for_formula(bottle_dep, inherited_options)
-          expanded_bottle_deps
-        end
-        expanded_deps = Dependency.merge_repeats(bottle_deps + expanded_deps)
-      end
-    end
+    expanded_deps = expand_dependencies_for_formula(formula, inherited_options)
```

Comment:
To be sure: is the intent here to just avoid handling `:all` bottles and treat them the same as binary bottles which pick up GCC/Glibc dependencies when the host's is too old?

If so, works for me 👍

This is the part that I previously said I don't fully follow yet, so, unless I misunderstood the above, I'm happy with this PR now.

Reply:
Yes, we're avoiding the special "don't install bottle dependencies for `:all` bottles" logic and just make GCC a build dependency for everything and glibc a runtime dependency for everything.

While this might end up with an extra or prematurely installed GCC or glibc for some users: I'm not convinced this optimisation is worth much given they are extremely likely to need it at some point anyway.

That said, I'm happy for someone else to do this just:
1) not me
2) not blocking this PR
3) not blocking the 22.04 migration

Reply:
> I'm not convinced this optimisation is worth much given they are extremely likely to need it at some point anyway.

Agreed. I'm also concerned that this is an optimisation that we never test, so could break at any point.

Reply:
> Agreed. I'm also concerned that this is an optimisation that we never test, so could break at any point.

Same here. I'm happy that this code is gone. Looks like premature optimisation for a niche use case. At the end of the day you'll always end up with glibc and gcc anyway on an old system.